`codegen_ssa` cleanups

nnethercote opened the pull request:
Some clarifications I made when reading this code closely.

r? @tmiasko

nnethercote committed:
Inline and remove `submit_pre_codegened_module_to_llvm`.

It has a single callsite, and provides little value.

nnethercote committed:
Add comments to `WorkItemResult`.

And rename the `Compiled` variant as `Finished`, because that name makes
it clearer there is nothing left to do, contrasting nicely with the
`Needs*` variants.

nnethercote committed:
Add some assertions.

- Thin and fat LTO can't happen together.
- `NeedsLink` and (non-allocator) `Compiled` work item results can't
  happen together.

nnethercote committed:
Rename things related to the main thread's operations.

It took me some time to understand how the main thread can lend a
jobserver token to an LLVM thread. This commit renames a couple of
things to make it clearer.

- Rename the `LLVMing` variant as `Lending`, because that is a clearer
  description of what is happening.
- Rename `running` as `running_with_own_token`, which makes it clearer
  that there might be one additional LLVM thread running (with a loaned
  token). Also add a comment to its definition.

nnethercote committed:
Rename `MainThreadWorkerState`.

The `Worker` is unnecessary, and just makes it longer than necessary.

nnethercote committed:
Remove an unnecessary `pub`.

nnethercote committed:
Remove some unused values in `codegen_crate`.

lqd commented on compiler/rustc_codegen_ssa/src/back/write.rs:
```diff
@@ -751,9 +751,19 @@ impl<B: WriteBackendMethods> WorkItem<B> {
     NeedsFatLTO(FatLTOInput<B>),
+
+    /// The backend has finished compiling a CGU, which now needs to go through
+    /// fat LTO.
```
small typo ?

```suggestion
    /// thin LTO.
```

nnethercote replied:
Fixed, thanks.

nnethercote committed:
Give the coordinator thread a name.

This is useful when profiling with a profiler like Samply.

nnethercote committed:
Remove `ExtraBackendMethods::spawn_thread`.

It's no longer used, and `spawn_named_thread` is preferable, because
naming threads is helpful when profiling.

nnethercote committed:
Fix a comment.

Make it match the corresponding comment at the start of the unstable
options.

nnethercote committed:
Document `-Zno-parallel-llvm`.

Because it's usefulness wasn't clear to me, and I initially wondered if
it could be removed. The text is based on the text in #50972, the PR
that added the flag.

nnethercote committed:
Move `maybe_start_llvm_timer`'s body into `spawn_work`.

The two functions are alway called together. This commit factors out the
repeated code.

nnethercote committed:
Remove `CodegenContext::worker`.

`CodegenContext` is immutable except for the `worker` field - we clone
`CodegenContext` in multiple places, changing the `worker` field each
time. It's simpler to move the `worker` field out of `CodegenContext`.

nnethercote committed:
Tweak a loop condition.

This loop condition involves `codegen_state`, `work_items`, and
`running_with_own_token`. But the body of the loop cannot modify
`codegen_state`, so repeatedly checking it is unnecessary.

nnethercote committed:
Tweak structure of the message loop.

The main loop has a *very* complex condition, which includes two
mentions of `codegen_state`. The body of the loop then immediately
switches on the `codegen_state`.

I find it easier to understand if it's a `loop` and we check for exit
conditions after switching on `codegen_state`. We end up with a tiny bit
of code duplication, but it's clear that (a) we never exit in the
`Ongoing` case, (b) we exit in the `Completed` state only if several
things are true (and there's interaction with LTO there), and (c) we
exit in the `Aborted` state if a couple of things are true. Also, the
exit conditions are all simple conjunctions.

nnethercote committed:
Use standard Rust capitalization rules for names containing "LTO".

nnethercote commented:
I added some more commits. Best reviewed one commit at a time.

nnethercote committed:
Introduce `running_with_any_token` closure.

It makes things a little clearer.

nnethercote committed:
Fix LLVM thread names on Windows.

PR #112946 tweaked the naming of LLVM threads, but messed things up
slightly, resulting in threads on Windows having names like `optimize
module {} regex.f10ba03eb5ec7975-cgu.0`.

This commit removes the extraneous `{} `.

nnethercote commented:
I added one more small commit. That should be the end of them :)

nnethercote committed:
Clean up `generate_lto_work`.

This function has some shared code for the thin LTO and fat LTO cases,
but those cases have so little in common that it's actually clearer to
treat them fully separately.

bjorn3 commented on compiler/rustc_codegen_ssa/src/base.rs:
```diff
@@ -790,7 +788,6 @@ pub fn codegen_crate<B: ExtraBackendMethods>(
                 );
-                true
             }
         };
```
This `;` can be removed, right?

nnethercote replied:
Yes. I'm surprised rustfmt didn't do that automatically. I have updated the code.

bjorn3 commented on compiler/rustc_codegen_ssa/src/back/write.rs:
```diff
@@ -1300,12 +1300,16 @@ fn start_executing_work<B: ExtraBackendMethods>(
 
+        // How many LLVM worker threads are running in total. This *includes*
+        // any that the main thread is lending a Token to.
+        let running_with_any_token = |main_thread_state, running_with_own_token| {
+            running_with_own_token
+                + if main_thread_state == MainThreadState::Lending { 1 } else { 0 }
+        };
```
I think this function should return a bool. It is always compared against 0 and "any" is an operation which normally returns a bool.

nnethercote replied:
I considered that, and I felt that the `running_with_own_token`/`running_with_any_token` distinction was better because it shows they are closely related. But I could be convinced otherwise.

bors commented:
:umbrella: The latest upstream changes (presumably #114264) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote committed:
Remove unnecessary semicolon.

nnethercote commented:
I have rebased.

bjorn3 commented:
@bors r+

bors commented:
:pushpin: Commit c17c8dc78e475e2df377b3cd5c56e6233a1a5157 has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 c17c8dc78e475e2df377b3cd5c56e6233a1a5157 -->
<!-- homu: {"type":"Approved","sha":"c17c8dc78e475e2df377b3cd5c56e6233a1a5157","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit c17c8dc78e475e2df377b3cd5c56e6233a1a5157 with merge 5082281609a7c3bcba0292b0db6a0d3e20bf9549...
<!-- homu: {"type":"BuildStarted","head_sha":"c17c8dc78e475e2df377b3cd5c56e6233a1a5157","merge_sha":"5082281609a7c3bcba0292b0db6a0d3e20bf9549"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5712324288/job/15478678731)
Approved by: bjorn3
Pushing 5082281609a7c3bcba0292b0db6a0d3e20bf9549 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"bjorn3","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5712324288/job/15478678731"},"merge_sha":"5082281609a7c3bcba0292b0db6a0d3e20bf9549"} -->

rust-timer commented:
Finished benchmarking commit (5082281609a7c3bcba0292b0db6a0d3e20bf9549): [comparison URL](https://perf.rust-lang.org/compare.html?start=03b8b50fcb0c95e1f5ce303572ed4516a59a9b15&end=5082281609a7c3bcba0292b0db6a0d3e20bf9549&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=03b8b50fcb0c95e1f5ce303572ed4516a59a9b15&end=5082281609a7c3bcba0292b0db6a0d3e20bf9549&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.7% | [0.7%, 0.7%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=03b8b50fcb0c95e1f5ce303572ed4516a59a9b15&end=5082281609a7c3bcba0292b0db6a0d3e20bf9549&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=03b8b50fcb0c95e1f5ce303572ed4516a59a9b15&end=5082281609a7c3bcba0292b0db6a0d3e20bf9549&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=03b8b50fcb0c95e1f5ce303572ed4516a59a9b15&end=5082281609a7c3bcba0292b0db6a0d3e20bf9549&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 652.303s -> 652.574s (0.04%)
